cpp/src/aztec/plonk/proof_system/prover/prover.cpp

```diff
@@ -285,30 +291,31 @@ template <typename settings> void ProverBase<settings>::execute_second_round()
         add_plookup_memory_records_to_w_4();
         std::string wire_tag = "w_4";
         barretenberg::polynomial& w_4_lagrange = key->polynomial_cache.get(wire_tag + "_lagrange");
-        // barretenberg::polynomial& wire_fft = key->polynomial_cache.get(wire_tag + "_fft");
-        barretenberg::polynomial w_4(key->circuit_size);
-        // barretenberg::polynomial_arithmetic::copy_polynomial(&wire[0], &wire_fft[0], n, n);
-        barretenberg::polynomial_arithmetic::copy_polynomial(&w_4_lagrange[0], &w_4[0], circuit_size, circuit_size);
 
-        // TODO: This adds blinding to the what will become the w_4_monomial, NOT to the w_4_lagrange poly. Is this
-        // intentional?
+        // add randomness to w_4_lagrange
         const size_t w_randomness = 3;
         ASSERT(w_randomness < settings::num_roots_cut_out_of_vanishing_polynomial);
         for (size_t k = 0; k < w_randomness; ++k) {
             // Blinding
-            w_4.at(circuit_size - settings::num_roots_cut_out_of_vanishing_polynomial + k) = fr::random_element();
+            w_4_lagrange.at(circuit_size - settings::num_roots_cut_out_of_vanishing_polynomial + k) =
+                fr::random_element();
         }
 
-        // poly w_4 and add to cache
-        w_4.ifft(key->small_domain);
-        key->polynomial_cache.put(wire_tag, std::move(w_4));
-
-        // Commit to w_4:
+        // commit to w_4 using Lagrange srs.
         queue.add_to_queue({
             .work_type = work_queue::WorkType::SCALAR_MULTIPLICATION,
-            .mul_scalars = key->polynomial_cache.get(wire_tag).get_coefficients(),
+            .mul_scalars = w_4_lagrange.get_coefficients(),
             .tag = "W_4",
-            .constant = barretenberg::fr(0),
+            .constant = work_queue::MSMType::LAGRANGE_N,
+            .index = 0,
+        });
+
+        // convert w_4 to the coefficient form.
```

Comment:
Was wondering why this was not previously done in the queue. Just an oversight I guess?

Reply:
I believe it wasn't an oversight. Previously we were committing to $w_4$ using the monomial SRS. So what we did is: 
1. $w_{4, \textsf{mono}} = \textsf{iFFT}(w_{4, \textsf{lag}})$  // this happens right away
2. $W_4 = \textsf{commit}(w_{4, \textsf{mono}})$  // this is pushed to the queue and is processed after the round
These two steps happens sequentially. So step 1 can be computed "locally" without having to wait for the queue processing at the end of this round. We could do the $\textsf{iFFT}$ operation on the queue but that would only work iff the iFFT and MSM are executed sequentially in WASM (so our typescript code would need to have this knowledge). 

Now, when we want to commit using the Lagrange SRS, we need to two operations:
1. $w_{4, \textsf{mono}} = \textsf{iFFT}(w_{4, \textsf{lag}})$  
2. $W_4 = \textsf{commit}(w_{4, \textsf{lag}})$ 
These two operations are computing the output on the same input, i.e. $w_{4, \textsf{lag}}$. If we did the iFFT "locally" and not on queue, that would change the polynomial $w_4$ to monomial form, and hence the MSM operation would give a wrong result (Lagrange SRS with coefficients of $w_4$). Therefore, in this case, there is no choice but to use the queue to compute iFFT. One downside of this though is: we will have to ensure that the MSM operation is processed before the iFFT operation in WASM (because iFFT modifies the same memory).